src/devices/GoPiGo3/GoPiGo3.cs

```diff
@@ -0,0 +1,900 @@
+        public double MotorTicksPerDegree => (MotorGearRatio * EncoderTicksPerRotation / 360.0);
+
+        /// <summary>
+        /// The 2 Groove Sensor
+        /// </summary>
+        public GrooveSensor[] GrooveSensor { get; internal set; }
+
+        /// <summary>
+        /// The GoPiGo3 information including hardware, firmware, ID and Manufacturer
+        /// </summary>
+        public GoPiGoInfo GoPiGo3Info { get; internal set; }
+
+        /// <summary>
+        /// Get the real 5V and Battery/VCC voltages
+        /// </summary>
+        public GoPiGoVoltage GoPiGoVoltage => new GoPiGoVoltage() { Voltage5V = GetVoltage5V(), VoltageBattery = GetVoltageBatteryVcc() };
+
+        #endregion
+
+        #region Constructor and Dispose
+
+        /// <summary>
+        /// Create a GoPiGo class
+        /// </summary>
+        /// <param name="spiAddress">The SPI address, by default 8</param>
+        /// <param name="autoDetect">Try to autodetect the board</param>
+        /// <param name="busId">SPI bus ID, by default 0 on a RaspberryPi</param>
+        /// <param name="chipSelectLine">SPI Chip Select Line, by default 1 on a Raspberryi</param>
+        public GoPiGo(byte spiAddress = 8, bool autoDetect = true, int busId = 0, int chipSelectLine = 1)
+        {
+            SpiAddress = spiAddress;
+            // SPI 0 is used on Raspberry with ChipSelectLine 1
+            var settings = new SpiConnectionSettings(busId, chipSelectLine);
+            // 500K is the SPI communication with GoPiGo
+            settings.ClockFrequency = 500000;
+            // see http://tightdev.net/SpiDev_Doc.pdf
+            settings.Mode = SpiMode.Mode0;
+            settings.DataBitLength = 8;
+            // as the SPI is a static, checking if it has already be initialised
+            if (_spiDevice == null)
+            {
+                _spiDevice = new UnixSpiDevice(settings);
+            }
+            GoPiGo3Info = new GoPiGoInfo();
+            if (autoDetect == true)
+                try
+                {
+                    GoPiGo3Info.Manufacturer = GetManufacturer();
+                    GoPiGo3Info.Board = GetBoard();
+                    GoPiGo3Info.SoftwareVersion = GetFirmwareVersion();
+                    GoPiGo3Info.HardwareVersion = GetHardwareVersion();
+                    GoPiGo3Info.Id = GetId();
+                }
+                catch (IOException)
+                {
+                    throw new IOException($"No SPI response. GoPiGo3 with address {SpiAddress} not connected.");
+                }
+            if ((GoPiGo3Info.Manufacturer != ManufacturerName) || (GoPiGo3Info.Board != BoardName))
+                throw new IOException($"GoPiGo3 with address {SpiAddress} not connected.");
+            MotorGearRatio = DefaultMotorGearRatio;
+            EncoderTicksPerRotation = DefaultEncoderTicksPerRotation;
+            // Initialise the 2 Groove sensors
+            GrooveSensor = new GrooveSensor[2];
```

Comment:
This can be

```csharp
new GrooveSensor[2]
{
  new GrooveSensor(GroovePort.Groove1),
  new GrooveSensor(GroovePort.Groove2)
}
```

Reply:
will do

Reply:
GrooveSensor?  The technical term is Grove though. 
https://www.seeedstudio.com/category/Grove-c-1003.html